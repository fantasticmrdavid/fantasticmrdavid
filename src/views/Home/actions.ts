import { HOME_MEDIA_IMAGE_UPDATED } from 'reducers/homeMedia';
import { MEDIA_LOADING_STOPPED } from 'reducers/loading';

export const homeMediaImageUpdated = (image: any) => ({
  type: HOME_MEDIA_IMAGE_UPDATED,
  image,
});

export const mediaLoadingStopped = () => ({
  type: MEDIA_LOADING_STOPPED,
});
