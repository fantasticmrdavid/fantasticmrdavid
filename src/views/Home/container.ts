import { connect } from 'react-redux';
import { AppState } from 'reducers';
import { DispatchProps, StateProps } from './types';
import Ui from './ui';

const mapStateToProps = (state: AppState) => {
  return {
    loading: state.loading.media,
  };
};

export default connect<StateProps, DispatchProps, {}, AppState>(mapStateToProps)(Ui);
