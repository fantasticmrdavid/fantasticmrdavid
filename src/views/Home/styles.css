.HomeMedia {
  background-color: $c-text-primary;
  background-position: center center;
  background-repeat: no-repeat;
  background-size: cover;
  width: 100%;
  height: 100vh;
  opacity: 1;
  transition: 1s;

  &.loading {
    opacity: 0;
  }
}
