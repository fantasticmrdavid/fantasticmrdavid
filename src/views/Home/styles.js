import styled from 'styled-components';
import theme from 'styles/theme';

export const Container = styled.div`
  background-image: ${props => `url(${props.imgSrc})`};
  background-color: ${theme.colorLightGrey};
  background-position: center center;
  background-repeat: no-repeat;
  background-size: cover;
  width: 100%;
  height: 100vh;
  opacity: ${props => (props.loading ? 0 : 1)};
  transition: 1s;
`;
