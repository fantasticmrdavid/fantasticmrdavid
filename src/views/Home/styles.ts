import styled from 'styled-components';
import theme from 'styles/theme';

interface ContainerProps {
  imgSrc: string,
  loading: boolean,
}

export const Container = styled.div<ContainerProps>`
  background-image: ${props => `url(${props.imgSrc})`};
  background-color: ${theme.colorLightGrey};
  background-position: center center;
  background-repeat: no-repeat;
  background-size: cover;
  width: 100%;
  height: 100vh;
  opacity: ${props => (props.loading ? 0 : 1)};
  transition: 1s;
`;
