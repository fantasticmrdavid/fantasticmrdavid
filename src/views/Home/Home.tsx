import React, { memo, useEffect, useState } from "react";
import { Container } from './home.styles';

interface HomeProps {
  isLoading: boolean
}

const fullImgSrc = "/assets/images/legoFlinders.jpg"

export const Home: React.FC<HomeProps> = memo(({ isLoading }) => {
  const [imgSrc, setImgSrc] = useState("/assets/images/legoFlinders_low.jpg")
  useEffect(() => {
    const img = new Image();
    img.src = fullImgSrc;
    img.onload = () => {
      setImgSrc(fullImgSrc);
    };
  }, [fullImgSrc])
  return <Container isLoading={isLoading} imgSrc={imgSrc} />;
});
