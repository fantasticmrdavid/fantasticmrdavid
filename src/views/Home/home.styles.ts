import styled, { css } from 'styled-components';
import { fadeIn } from 'styles/animations';
import theme from 'styles/theme';

interface ContainerProps {
  imgSrc: string,
  isLoading: boolean,
}

export const Container = styled.div<ContainerProps>`
  background-image: ${(props) => `url(${props.imgSrc})`};
  background-color: ${theme.colorLightGrey};
  background-position: center center;
  background-repeat: no-repeat;
  background-size: cover;
  width: 100%;
  height: 100vh;
  opacity: ${(props) => (props.isLoading ? 0 : 1)};
  transition: 1s;
  animation: ${css`${fadeIn} 1s ease 1`};
  animation-fill-mode: forwards;
`;
