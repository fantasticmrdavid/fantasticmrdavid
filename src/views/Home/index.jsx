import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { homeMediaImageUpdated, mediaLoadingStopped } from 'actions';
import * as styles from './styles';

const HOME_IMAGES = [
  {
    name: 'legoFlinders',
    src: '/assets/images/legoFlinders.jpg',
  },
];

class Home extends PureComponent {
  componentWillMount() {
    const { updateImage } = this.props;
    const image = HOME_IMAGES[Math.floor(Math.random() * HOME_IMAGES.length)];
    updateImage(image);
  }

  componentDidMount() {
    const { image, stopLoading } = this.props;
    const bg = new Image();
    bg.onload = () => {
      stopLoading();
    };
    bg.src = image.src;

    return bg;
  }

  render() {
    const { loading, image } = this.props;
    const { Container } = styles;
    return <Container loading={loading} imgSrc={image.src} />;
  }
}

const mapStateToProps = (state) => {
  return {
    image: state.homeMedia.image,
    loading: state.loading.media,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    updateImage: (image) => {
      dispatch(homeMediaImageUpdated(image));
    },
    stopLoading: () => {
      dispatch(mediaLoadingStopped());
    },
  };
};

Home.propTypes = {
  updateImage: PropTypes.func.isRequired,
  stopLoading: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  image: PropTypes.object,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Home);
