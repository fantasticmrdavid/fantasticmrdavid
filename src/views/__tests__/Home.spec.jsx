import React from 'react';
import { Provider } from 'react-redux';
import store from 'store';
import { mount } from 'enzyme';
import Initializer from 'routers/Initializer';
import Home from 'views/Home';
import Ui from 'views/Home/ui';

describe('View -> Home', () => {
  it('should render the Home view with Initializer without throwing an error', () => {
    expect(mount(
      <Provider store={store}>
        <Initializer />
        <Home />
      </Provider>,
    )).toMatchSnapshot();
  });

  it('should render the Home UI (loading) without throwing an error', () => {
    const props = {
      image: {
        name: 'test',
        src: 'testImage/url.jpg',
      },
      loading: true,
      updateImage: jest.fn(),
    };
    expect(mount(<Ui {...props} />)).toMatchSnapshot();
  });

  it('should render the Home UI (not loading) without throwing an error', () => {
    const props = {
      image: {
        name: 'test',
        src: 'testImage/url.jpg',
      },
      loading: false,
      updateImage: jest.fn(),
    };
    expect(mount(<Ui {...props} />)).toMatchSnapshot();
  });
});
