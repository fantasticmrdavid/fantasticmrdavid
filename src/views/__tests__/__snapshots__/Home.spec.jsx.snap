// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View -> Home should render the Home UI (loading) without throwing an error 1`] = `
<Memo()
  image={
    Object {
      "name": "test",
      "src": "testImage/url.jpg",
    }
  }
  loading={true}
  updateImage={[MockFunction]}
>
  <styles__Container
    imgSrc="/assets/images/legoFlinders.jpg"
    isLoading={false}
  >
    <div
      className="styles__Container-sc-19v7chj-0 ftUPBY"
    />
  </styles__Container>
</Memo()>
`;

exports[`View -> Home should render the Home UI (not loading) without throwing an error 1`] = `
<Memo()
  image={
    Object {
      "name": "test",
      "src": "testImage/url.jpg",
    }
  }
  loading={false}
  updateImage={[MockFunction]}
>
  <styles__Container
    imgSrc="/assets/images/legoFlinders.jpg"
    isLoading={false}
  >
    <div
      className="styles__Container-sc-19v7chj-0 ftUPBY"
    />
  </styles__Container>
</Memo()>
`;

exports[`View -> Home should render the Home view with Initializer without throwing an error 1`] = `
<Memo()>
  <styles__Container
    imgSrc="/assets/images/legoFlinders.jpg"
    isLoading={false}
  >
    <div
      className="styles__Container-sc-19v7chj-0 ftUPBY"
    />
  </styles__Container>
</Memo()>
`;
